src/common/fault_domain.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * (C) Copyright 2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#ifndef __DAOS_COMMON_FAULT_DOMAIN__
+#define __DAOS_COMMON_FAULT_DOMAIN__
+
+#include <daos/common.h>
+
+/**
+ * Possible types of fault domain tree node.
+ */
+enum d_fd_node_type {
+	D_FD_NODE_TYPE_UNKNOWN = 0,
+	D_FD_NODE_TYPE_DOMAIN,
+	D_FD_NODE_TYPE_RANK,
+};
+
+/**
+ * Unified structure representing a node in the fault domain tree.
+ *
+ */
+struct d_fd_node {
+	enum d_fd_node_type fdn_type; /* indicates which union member to use */
+	union {
+		const struct d_fault_domain		*dom;
+		const struct d_fault_domain_rank	*rank;
+	} fdn_val;
+};
+
+/**
+ * Represents a fault domain at any level above rank.
+ */
+struct d_fault_domain {
+	uint32_t fd_level;	/** level in the fault domain tree >= 1 */
+	uint32_t fd_id;		/** unique ID */
+	uint32_t fd_children_nr; /** number of children */
+};
+
+/** Represents a fault domain rank */
+struct d_fault_domain_rank {
+	uint32_t fdr_rank; /** rank number */
+};
```

Comment:
I was thinking it might be useful to have a struct in case we decide we need to change the format to provide more info at the rank level (like an ID that's tree-unique, instead of level-unique). I don't think that's necessary at this point, though.